File: client/src/shared/lib/honor.ts
// Honor tiers — display math only.
//
// MUST stay in sync with the server: server/internal/user/honor.go (the tier
// bands and the token strings). This is the same manual-sync convention as
// xpLevel.ts ↔ level.go and wsEvents.ts ↔ events.go — there is no generated
// shared type.
//
// The SERVER is authoritative for the score AND the tier: both arrive on the
// user object, the profile response, and event:honor_updated. Nothing here ever
// makes a gating decision (Story 9.8's room gate is enforced server-side). Keep
// this the ONLY client copy of the honor bands (Story 9.7 Design Decision D6).

/** The five stable tier tokens the server emits, in descending order. */
export const HONOR_TIERS = ["exemplary", "trusted", "fair", "unreliable", "problematic"] as const;

export type HonorTier = (typeof HONOR_TIERS)[number];

/** Inclusive lower bound of each tier band. Mirrors HonorTier in honor.go. */
const HONOR_TIER_FLOORS: ReadonlyArray<readonly [HonorTier, number]> = [
  ["exemplary", 95],
  ["trusted", 85],
  ["fair", 70],
  ["unreliable", 50],
  ["problematic", 0],
];

/**
 * The score a player with no history scores — the Beta(4,1) prior,
 * 100 * 4 / 5. Used as the render fallback when a score is somehow absent, so
 * the UI never shows a misleading 0 ("Problematic") for a blank slate.
 *
 * Wire it through `honorScoreOrPrior` — do NOT re-implement the check with `||`,
 * which would swallow a real score of 0.
 */
export const HONOR_PRIOR_SCORE = 80;

/**
 * How many matches each side of the trend comparison spans. Mirrors
 * `honorTrendWindow` in server/internal/user/honor.go — same manual-sync rule as
 * the tier bands above.
 *
 * Display only: it exists so the `profile.honor.trendTooltip` copy can interpolate
 * the real number instead of hardcoding "20" in four locale files, which is how
 * the tooltip would otherwise drift the moment the server constant is retuned
 * (review pass 2). It never feeds a calculation — the server computes the trend.
 */
export const HONOR_TREND_WINDOW = 20;

/**
 * How many FINISHED-OR-ABANDONED matches an account needs before it earns a
 * score. Mirrors `honorNewPlayerMinMatches` in server/internal/user/honor.go —
 * same manual-sync rule as the tier bands and the trend window above.
 *
 * Display only, and only ever as a denominator: it renders the newcomer's
 * "2 / 5" progress so the New Player state says how to LEAVE it, instead of
 * being a label that explains nothing. The server owns the decision — always
 * branch on its `isNewPlayer` flag, never on `completed + abandoned < 5`
 * computed here. That floor counts experience, not successes, and re-deriving it
 * client-side is exactly the bypass two review passes closed on the server.
 */
export const HONOR_NEW_PLAYER_MIN_MATCHES = 5;

/**
 * Coerce a possibly-absent honor score into a renderable number.
 *
 * The WS handler type-guards every field before it reaches the store, but the
 * HTTP paths do not: a client bundle newer than the server (or a server rollback
 * mid-deploy) yields `undefined` here. Left unguarded that became `NaN`, every
 * tier-floor comparison went false, and the UI fell through to the WORST tier —
 * a blank number in danger red with `data-tier="problematic"`, which is the exact
 * misread HONOR_PRIOR_SCORE was documented to prevent (code review 2026-07-29).
 *
 * Uses Number.isFinite, never truthiness: a score of 0 is a REAL value from Go
 * and must survive untouched.
 */
export function honorScoreOrPrior(score: number | null | undefined): number {
  return Number.isFinite(score) ? (score as number) : HONOR_PRIOR_SCORE;
}

/**
 * Bucket a server-supplied score into its tier. Prefer the server's own
 * `honorTier` when you have it; use this only where a score arrives without one
 * (e.g. rendering a hypothetical or a fallback).
 *
 * A non-finite score falls back to the prior's band rather than "problematic".
 */
export function honorTierForScore(score: number): HonorTier {
  const clamped = Math.min(100, Math.max(0, honorScoreOrPrior(score)));
  for (const [tier, floor] of HONOR_TIER_FLOORS) {
    if (clamped >= floor) return tier;
  }
  return "problematic";
}

/**
 * Narrow an arbitrary server string to a known tier, falling back to the
 * score's own bucket for an unrecognised token.
 *
 * This is the version-skew guard: a future server tier retune must degrade to a
 * sensible colour on a stale bundle rather than rendering a missing i18n key.
 * It is why the Zod payload schema types honorTier as a plain string.
 */
export function normalizeHonorTier(tier: string, score: number): HonorTier {
  return (HONOR_TIERS as readonly string[]).includes(tier)
    ? (tier as HonorTier)
    : honorTierForScore(score);
}

/** Bar fill in [0, 1] for an honor meter. The scale is a flat 0-100. */
export function honorBarFill(score: number): number {
  return Math.min(1, Math.max(0, honorScoreOrPrior(score) / 100));
}

/**
 * Coerce a possibly-absent honor count or trend delta into a renderable integer.
 *
 * Same threat model as `honorScoreOrPrior` — the HTTP profile and auth-refresh
 * responses are not type-guarded the way the WS payload is, so a bundle newer
 * than the server yields `undefined`. Left raw those rendered a blank number and
 * `data-value="undefined"` (review pass 2). Uses Number.isFinite, so a real 0
 * survives.
 */
export function honorCountOrZero(value: number | null | undefined): number {
  return Number.isFinite(value) ? (value as number) : 0;
}

/**
 * Coerce a possibly-absent `isNewPlayer` flag, defaulting to SUPPRESSED.
 *
 * The safe default is `true`: if we cannot tell whether this account has enough
 * history to have earned a score, hiding the number is the conservative choice.
 * Left raw, `undefined` was falsy and therefore took the NUMERIC branch, so a
 * server that had not yet shipped the honor fields made every account — brand-new
 * ones included — render a confident "80 / Fair" (review pass 2).
 */
export function honorIsNewPlayer(value: boolean | null | undefined): boolean {
  return typeof value === "boolean" ? value : true;
}

/**
 * The five bands as half-open [from, to) ranges over 0-100, ascending.
 *
 * DERIVED from HONOR_TIER_FLOORS rather than restated, so the banded meter's
 * gradient stops, its tick labels and the tier a score buckets into can never
 * disagree. Retuning a floor moves all three at once.
 */
const HONOR_FLOORS_ASCENDING = [...HONOR_TIER_FLOORS].reverse();

export const HONOR_TIER_BANDS: ReadonlyArray<{ tier: HonorTier; from: number; to: number }> =
  HONOR_FLOORS_ASCENDING.map(([tier, from], i) => {
    const next = HONOR_FLOORS_ASCENDING[i + 1];
    return { tier, from, to: next ? next[1] : 100 };
  });

/**
 * Tier → colour, the ONLY copy. Values are `var()` references into the honour
 * ramp declared in index.css (`--h1`…`--h5`, worst→best), so they re-root
 * automatically inside `.game-table` and every honour surface themes itself on
 * felt with no fork.
 *
 * Centralised here for the reason coinGold.ts states for COIN_GOLD: before this,
 * the map was duplicated in HonorPanel.tsx and TopBar.tsx and the two had
 * ALREADY drifted — they disagreed on `fair` — with nothing in TypeScript able
 * to notice. The redesign adds five more consumers (the room badge, the seat
 * shield, the ejection meter, the slider fill, the reconnect line), so the
 * duplication had to stop before it multiplied.
 *
 * Prefer `<HonorShield>` (shared/components/HonorShield.tsx) over reading these
 * directly: it pairs the colour with the tier's GLYPH, which is what keeps the
 * scale legible without relying on colour alone.
 */
export const HONOR_TIER_COLOR: Record<HonorTier, string> = {
  exemplary: "var(--h5)",
  trusted: "var(--h4)",
  fair: "var(--h3)",
  unreliable: "var(--h2)",
  problematic: "var(--h1)",
};

/** Tier → low-alpha fill, for badge/chip grounds and meter bands. */
export const HONOR_TIER_SOFT: Record<HonorTier, string> = {
  exemplary: "var(--h5-soft)",
  trusted: "var(--h4-soft)",
  fair: "var(--h3-soft)",
  unreliable: "var(--h2-soft)",
  problematic: "var(--h1-soft)",
};

/** Tier → hairline/border tone, for outlined variants. */
export const HONOR_TIER_LINE: Record<HonorTier, string> = {
  exemplary: "var(--h5-line)",
  trusted: "var(--h4-line)",
  fair: "var(--h3-line)",
  unreliable: "var(--h2-line)",
  problematic: "var(--h1-line)",
};

/** The subset of a room the honour gate reads. Structural, so both the API
 *  `Room` type and a hand-built preview object satisfy it. */
export type HonorGateRoom = {
  minHonor?: number | null;
  allowNewPlayers?: boolean | null;
};

/** The subset of the viewer the honour gate reads. */
export type HonorGateViewer = {
  honorScore?: number | null;
  isNewPlayer?: boolean | null;
};

/**
 * Does this room enforce an honour requirement at all? Mirrors
 * `(*Room).honorGated()` in server/internal/room/handler.go.
 *
 * Absent fields read as UNGATED, which matters in practice: the QuickPlay
 * `system:room_created` payload is hand-built and omits both keys, and a
 * synthesized quick-play room genuinely is ungated.
 */
export function honorRoomIsGated(room: HonorGateRoom): boolean {
  return honorCountOrZero(room.minHonor) > 0 || room.allowNewPlayers === false;
}

/**
 * Would this viewer pass this room's honour gate? Mirrors `honorGateError` in
 * server/internal/room/handler.go — isNewPlayer is evaluated FIRST and a New
 * Player is never score-checked, which is the counter-intuitive half of Story
 * 9.8 D1 (the toggle is the owner's explicit "I'll take an unknown" switch).
 *
 * COSMETIC ONLY. The server re-validates on every join, return and start; this
 * exists so the lobby can render a Locked button and an "I qualify" filter
 * instead of making the player discover the gate by clicking. Never use it to
 * suppress a request — send it and let the 409 be authoritative.
 *
 * Note the deliberate asymmetry with CreateRoomModal's `honorKnown` check: there
 * an unknown envelope must NOT deny (a capability gate turning "unknown" into
 * "denied" would lock a veteran out of their own room), whereas here an unknown
 * viewer resolves through `honorIsNewPlayer`'s suppressed default and so is
 * treated leniently. Both err toward letting the player try.
 */
export function honorQualifies(room: HonorGateRoom, viewer: HonorGateViewer): boolean {
  if (!honorRoomIsGated(room)) return true;
  if (honorIsNewPlayer(viewer.isNewPlayer)) return room.allowNewPlayers !== false;
  return honorScoreOrPrior(viewer.honorScore) >= honorCountOrZero(room.minHonor);
}

/** The three stable trend-direction tokens the server emits. */
export const HONOR_TREND_DIRECTIONS = ["up", "flat", "down"] as const;

export type HonorTrendDirection = (typeof HONOR_TREND_DIRECTIONS)[number];

/**
 * Narrow an arbitrary server string to a known trend direction, defaulting to
 * "flat" — the neutral reading, which is the safe thing to show when the token
 * is unrecognised.
 */
export function normalizeHonorTrendDirection(direction: string): HonorTrendDirection {
  return (HONOR_TREND_DIRECTIONS as readonly string[]).includes(direction)
    ? (direction as HonorTrendDirection)
    : "flat";
}

File: server/internal/user/honor.go
package user

import (
	"math"
	"time"
)

// Honor score (Story 9.7). Honor answers one question: "does this player finish
// the matches they start, RIGHT NOW?" It is a server-authoritative signal (NFR8)
// and THIS FILE IS ITS SINGLE SOURCE OF TRUTH.
//
// The client carries one documented mirror at client/src/shared/lib/honor.ts,
// under the same manual-sync convention as level.go / xpLevel.ts and
// events.go / wsEvents.ts. That mirror is DISPLAY ONLY — it buckets a
// server-supplied score into a tier for colouring and never makes a gating
// decision. If you change the formula or the tier bands here, change it there
// in the same commit.
//
// Two buckets only — completed and abandoned (Story 9.7 D1). "Rage quit" and
// "disconnect" are the same event on this server (a socket close), and per-move
// timer expiry auto-plays a card instead of abandoning anyone, so the epic's
// rage_quits / timeout_abandons buckets have no producer and are not modelled.
//
// Every function here is PURE: no DB, no clock reads (time is always a
// parameter), no side effects. That is what makes the honor math table-testable
// and what lets the same arithmetic run in the profile read path, on the auth
// envelope, inside event:honor_updated, and in Story 9.8's join gate without
// any of them disagreeing.

// Tunable constants. These are placeholders per
// sprint-change-proposal-2026-06-18.md#Section 4 ("honor weights stay as
// placeholders, tuned during each story's planning"), so they are named consts:
// a retuning pass must never have to touch logic.
const (
	// honorHalfLifeDays is how many days it takes for one match's contribution
	// to decay to half its weight. 90 days: an abandonment is half-forgiven in
	// three months and ~94% gone in a year.
	honorHalfLifeDays = 90.0

	// honorAbandonPenalty is how many completed matches one abandonment
	// offsets. At 4.0, a single abandon in twenty drops Trusted to Fair.
	honorAbandonPenalty = 4.0

	// honorPriorCompleted is the Bayesian pseudo-count of completed matches
	// (added to BOTH numerator and denominator). Beta(4,1) smoothing — the
	// standard remedy for "1-of-1 reads as 100%".
	honorPriorCompleted = 4.0

	// honorPriorAbandoned is the Bayesian pseudo-count of abandoned matches
	// (denominator only). Together with honorPriorCompleted it puts a player
	// with no history at 100*4/5 = 80.
	honorPriorAbandoned = 1.0

	// honorNewPlayerMinMatches is the raw FINISHED-OR-ABANDONED match count below
	// which the client suppresses the score and tier and shows a "New Player"
	// chip. Compared against the RAW lifetime totals, never the decayed weights.
	honorNewPlayerMinMatches = 5

	// honorTrendWindow is how many recent matches the trend comparison spans.
	honorTrendWindow = 20

	// honorTrendThreshold is the minimum |delta| between the recent window's
	// score and the PRECEDING window's score before the trend renders as up/down
	// rather than flat. Both sides are equal-size windows — see
	// HonorTrendWindowed for why comparing against the lifetime score was wrong.
	honorTrendThreshold = 2

	// honorEventWeight is what one just-finished match contributes to its
	// bucket. It is 1.0 by definition — decay is applied to the STORED weight
	// before this is added, so a brand-new event is always at full strength.
	honorEventWeight = 1.0

	// honorMaxClockSkew is how far ahead of `now` a stored honor_decayed_at may
	// be before ApplyHonorEvents stops trusting it and resets the reference to
	// now. A stamp in the future is normal at small magnitudes (clock skew
	// between app instances, an NTP step, or the 000017 backfill's Postgres
	// NOW() running marginally ahead of the app host) and must be clamped
	// forward, never rolled back — rolling it back double-decays the interval.
	// But clamping forward WITHOUT a ceiling means DecayFactor returns 1.0 for
	// as long as the stamp stays ahead, so a wildly wrong stamp would freeze
	// decay, and therefore freeze forgiveness, indefinitely and silently.
	// 24h is far beyond any plausible skew and far below any timescale the
	// 90-day half-life cares about.
	honorMaxClockSkew = 24 * time.Hour
)

// Tier tokens. The server returns these STABLE MACHINE TOKENS; the client maps
// token -> i18n label + colour. A display string must never cross the wire.
const (
	HonorTierExemplary   = "exemplary"   // 95-100
	HonorTierTrusted     = "trusted"     // 85-94
	HonorTierFair        = "fair"        // 70-84
	HonorTierUnreliable  = "unreliable"  // 50-69
	HonorTierProblematic = "problematic" // 0-49
)

// Trend direction tokens, same machine-token rule as the tiers.
const (
	HonorTrendUp   = "up"
	HonorTrendFlat = "flat"
	HonorTrendDown = "down"
)

// HonorEvent is one finished match's honor contribution for one player: the
// only thing that varies is which of the two buckets it lands in. A struct
// rather than a bare bool so the call sites read as data, not as a mystery
// boolean, and so a future third bucket (see Story 9.7 D1 — none exists today)
// is an additive change.
type HonorEvent struct {
	// Abandoned is true for the single seat whose reconnect window expired, and
	// false for every seat that saw the match through (including the three
	// innocent players in an abandoned match).
	Abandoned bool
}

// HonorSnapshot is one player's honor state immediately after a write. It is
// what event:honor_updated and the profile DTO render, so it carries the
// AUTHORITATIVE recomputed score rather than the lagging honor_score column.
type HonorSnapshot struct {
	Score          int
	Tier           string
	CompletedTotal int64
	AbandonedTotal int64
	IsNewPlayer    bool
	// Level is the player's lifetime level (Story 9.5 curve over total_xp),
	// carried here so the room roster can show level beside honor from the ONE
	// user read HonorForUsers already does. Populated ONLY by HonorForUsers —
	// NewHonorSnapshot has no user row and leaves it zero, so never render it
	// from a snapshot built anywhere else.
	Level int
}

// NewHonorSnapshot builds the renderable state from a user's stored honor
// columns, recomputing the score at `now`. This is the ONE place read paths
// (profile, auth envelope, match-end event) should go through, so none of them
// can accidentally reach for the stale honor_score column instead.
func NewHonorSnapshot(completedWeight, abandonedWeight float64, decayedAt *time.Time, completedTotal, abandonedTotal int64, now time.Time) HonorSnapshot {
	score := HonorScore(completedWeight, abandonedWeight, decayedAt, now)
	return HonorSnapshot{
		Score:          score,
		Tier:           HonorTier(score),
		CompletedTotal: completedTotal,
		AbandonedTotal: abandonedTotal,
		IsNewPlayer:    IsNewPlayer(completedTotal, abandonedTotal),
	}
}

// DecayFactor returns the multiplier that ages the stored honor weights forward
// from decayedAt to now: 0.5 ^ (elapsedDays / honorHalfLifeDays).
//
// A nil decayedAt means "never decayed" (the column is NULL) and returns exactly
// 1.0. A now that precedes decayedAt — clock skew, or a caller passing a
// deliberately fixed test clock — also returns 1.0 rather than a factor above 1,
// so weights can never be inflated by going backwards in time.
//
// This is the whole reason a running weight is exact rather than approximate:
// because every stored term decays by this same factor,
//
//	Σ 0.5^((now − tᵢ)/H) = 0.5^((now − last)/H) · Σ 0.5^((last − tᵢ)/H)
//
// so "decay forward, then add the new event" is algebraically identical to
// summing every match's weight from scratch (Story 9.7 D3).
func DecayFactor(decayedAt *time.Time, now time.Time) float64 {
	if decayedAt == nil {
		return 1.0
	}
	elapsed := now.Sub(*decayedAt)
	if elapsed <= 0 {
		return 1.0
	}
	days := elapsed.Hours() / 24.0
	return math.Pow(0.5, days/honorHalfLifeDays)
}

// HonorScore is the AUTHORITATIVE honor value: it recomputes from the stored
// weights on every read, so it is never stale. The users.honor_score column is a
// denormalized snapshot for SQL filtering only — never render it, never gate on
// it (see the migration 000017 header).
//
//	f     = DecayFactor(decayedAt, now)
//	C     = completedWeight × f
//	A     = abandonedWeight × f
//	honor = round( 100 × (C + 4) / (C + 4 + 4·A + 1) )   clamped to [0,100]
//
// Note the decay direction: an INACTIVE player's C and A both shrink, so their
// honor drifts back toward the 100*4/5 = 80 prior. That is intended — honor is a
// statement about current reliability, not a lifetime trophy.
//
// Negative weights are impossible (the DB CHECKs forbid them) but are clamped to
// zero anyway so a corrupt row degrades to the prior instead of producing
// nonsense. Rounding is math.Round (half away from zero), matching the Postgres
// ROUND(numeric) used by the migration backfill so both paths agree exactly.
func HonorScore(completedWeight, abandonedWeight float64, decayedAt *time.Time, now time.Time) int {
	f := DecayFactor(decayedAt, now)

	c := completedWeight * f
	a := abandonedWeight * f
	// Non-finite inputs degrade to the prior, never to a tier. +Inf is reachable
	// only by a manual/operator write (Postgres numeric accepts 'Infinity' and the
	// CHECK (>= 0) permits it), but left unguarded it produced Inf/Inf = NaN ->
	// math.Round(NaN) = NaN -> int(NaN) is implementation-defined (a large
	// negative on amd64) -> the score < 0 clamp turned it into 0, i.e. the WORST
	// tier in danger red. That was the opposite of how the tested NaN case
	// behaves. Both now land on the 80 prior. (Review pass 2.)
	if c < 0 || math.IsNaN(c) || math.IsInf(c, 0) {
		c = 0
	}
	if a < 0 || math.IsNaN(a) || math.IsInf(a, 0) {
		a = 0
	}

	numerator := 100 * (c + honorPriorCompleted)
	denominator := c + honorPriorCompleted + honorAbandonPenalty*a + honorPriorAbandoned
	// denominator is >= honorPriorCompleted + honorPriorAbandoned = 5 for any
	// non-negative c/a, so it can never be zero.
	score := int(math.Round(numerator / denominator))

	if score < 0 {
		score = 0
	}
	if score > 100 {
		score = 100
	}
	return score
}

// HonorScoreForCounts computes honor from raw undecayed match counts. Used for
// the recent-window trend, where the 20-match window IS the recency mechanism so
// no additional decay applies (Story 9.7 "Trend").
func HonorScoreForCounts(completed, abandoned int, now time.Time) int {
	return HonorScore(float64(completed), float64(abandoned), nil, now)
}

// HonorTier buckets a score into one of the five stable tokens. Scores outside
// [0,100] cannot occur (HonorScore clamps) but are bucketed defensively at the
// nearest end rather than returning an empty string, which would render as a
// missing i18n key.
func HonorTier(score int) string {
	switch {
	case score >= 95:
		return HonorTierExemplary
	case score >= 85:
		return HonorTierTrusted
	case score >= 70:
		return HonorTierFair
	case score >= 50:
		return HonorTierUnreliable
	default:
		return HonorTierProblematic
	}
}

// IsNewPlayer reports whether the client should suppress the numeric score and
// tier in favour of a "New Player" chip.
//
// It takes the RAW lifetime totals, never the decayed weights: a returning
// veteran's weights have decayed toward zero, and labelling them "New Player"
// would be both wrong and insulting.
//
// The floor counts EXPERIENCE (completed + abandoned), not successes. Flooring on
// completions alone — which is what AC2/AC6 literally specify and what shipped
// for review — let the worst possible actor hide behind the newcomer chip
// forever: 0 completed and 20 abandoned is a real score of 5 ("problematic"),
// yet it suppressed identically to a genuine first-timer, and Story 9.8's gate
// reads isNewPlayer off the same envelope. Overridden by PO decision 2026-07-29
// (code review), in the same spirit as D1-D3.
//
// Suppression is PRESENTATION ONLY. The server still returns the real score and
// tier when this is true, because Story 9.8's join gate needs them.
func IsNewPlayer(completedTotal, abandonedTotal int64) bool {
	return completedTotal+abandonedTotal < honorNewPlayerMinMatches
}

// HonorTrendWindow exposes how many recent matches the trend query should read,
// so the caller and the math cannot disagree about the window size.
func HonorTrendWindow() int {
	return honorTrendWindow
}

// HonorTrend compares two honor scores and returns the signed delta plus a
// stable direction token. A small delta reads as flat so ordinary noise does not
// render as a trend arrow.
//
// Both arguments MUST come from samples of the same size — see
// HonorTrendWindowed, which is the only caller that should exist.
func HonorTrend(recentScore, baselineScore int) (delta int, direction string) {
	delta = recentScore - baselineScore
	switch {
	case delta >= honorTrendThreshold:
		return delta, HonorTrendUp
	case delta <= -honorTrendThreshold:
		return delta, HonorTrendDown
	default:
		return delta, HonorTrendFlat
	}
}

// HonorTrendWindowed compares the newest window of matches against the window
// immediately before it, and is the ONLY correct way to compute the trend.
//
// It renders flat unless the two windows hold the SAME number of matches. That
// guard is the whole point of the two-window shape (code review 2026-07-29): the
// Beta(4,1) prior adds 4 pseudo-completions, and their drag depends on sample
// size. Comparing a 20-match window (a flawless one caps at 100*24/25 = 96)
// against a lifetime score computed from an unbounded decayed weight (98-100 for
// an active player) produced a permanent "Slipping -3" for players who had never
// abandoned a single match, and "Improving +12" for a player who had been idle
// for a year. Equal windows carry an identical prior drag, so the difference is
// behaviour rather than arithmetic.
//
// Consequence worth knowing: a player needs 2*honorTrendWindow finished matches
// before any trend renders. Below that the honest answer is "not enough
// evidence", and flat is how that is shown.
func HonorTrendWindowed(recentCompleted, recentAbandoned, priorCompleted, priorAbandoned int, now time.Time) (delta int, direction string) {
	recentTotal := recentCompleted + recentAbandoned
	priorTotal := priorCompleted + priorAbandoned
	if recentTotal == 0 || recentTotal != priorTotal {
		return 0, HonorTrendFlat
	}
	return HonorTrend(
		HonorScoreForCounts(recentCompleted, recentAbandoned, now),
		HonorScoreForCounts(priorCompleted, priorAbandoned, now),
	)
}
